src/mongoc/mongoc-uri.c

```diff
@@ -1073,3 +1117,102 @@ mongoc_uri_get_ssl (const mongoc_uri_t *uri) /* IN */
            BSON_ITER_HOLDS_BOOL (&iter) &&
            bson_iter_bool (&iter));
 }
+
+int32_t
+mongoc_uri_get_option_as_int32(const mongoc_uri_t *uri, const char *option,
+                               int32_t fallback)
+{
+   const bson_t *options;
+   bson_iter_t iter;
+   int32_t retval = fallback;
+
+   if ((options = mongoc_uri_get_options (uri)) &&
+        bson_iter_init_find_case (&iter, options, option) &&
+        BSON_ITER_HOLDS_INT32 (&iter)) {
+
+      if (!(retval = bson_iter_int32(&iter))) {
+         retval = fallback;
+      }
+   }
+
+   return retval;
+}
+
+bool
+mongoc_uri_set_option_as_int32(mongoc_uri_t *uri, const char *option,
+                               int32_t value)
+{
+   const bson_t *options;
+   bson_iter_t iter;
+
+   if (!mongoc_uri_option_is_int32 (option)) {
+      return false;
+   }
+
+   if ((options = mongoc_uri_get_options (uri)) &&
+         bson_iter_init_find_case (&iter, options, option)) {
+      if (BSON_ITER_HOLDS_INT32 (&iter)) {
+         bson_iter_overwrite_int32 (&iter, value);
+         return true;
+      } else {
+         return false;
+      }
+   }
+   bson_append_int32(&uri->options, option, -1, value);
+   return true;
+}
+
+bool
+mongoc_uri_set_option_as_bool(mongoc_uri_t *uri, const char *option,
+                              bool value)
+{
+   const bson_t *options;
+   bson_iter_t iter;
+
+   if (!mongoc_uri_option_is_bool (option)) {
+      return false;
+   }
+
+   if ((options = mongoc_uri_get_options (uri)) &&
+         bson_iter_init_find_case (&iter, options, option)) {
+      if (BSON_ITER_HOLDS_BOOL (&iter)) {
+         bson_iter_overwrite_bool (&iter, value);
+         return true;
+      } else {
+         return false;
+      }
+   }
+   bson_append_bool(&uri->options, option, -1, value);
+   return true;
+
+}
+
+bool
+mongoc_uri_set_option_as_utf8(mongoc_uri_t *uri, const char *option,
+                              const char *value)
+{
+   const bson_t *options;
+   bson_iter_t iter;
+   size_t len;
+
+   len = strlen(value);
+
+   if (!bson_utf8_validate(value, len, false)) {
+      return false;
+   }
+
+   if (!mongoc_uri_option_is_utf8 (option)) {
+      return false;
+   }
+
+   if ((options = mongoc_uri_get_options (uri)) &&
+       bson_iter_init_find_case (&iter, options, option)) {
+      return false;
```

Comment:
Yeah, let's do it. Test it well. It's not a performance issue since this is in configuration code. I suggest iterate the source bson key-by-key, copying values to the destination bson, until you reach "option", then write "value" instead. bson_iter_value and bson_append_value will allow you to ignore the types until you reach a key with the name "option". If you don't find that key, put option and value at the end.

My argument is that we shouldn't allow users to overwrite ints and bools, but prohibit them from overwriting strings, just because one is easy and the other is hard. That's an implementation detail.